programs/manifest/src/state/global.rs

```diff
@@ -236,36 +299,151 @@ impl<Fixed: DerefOrBorrowMut<GlobalFixed>, Dynamic: DerefOrBorrowMut<[u8]>>
+        let global_trader: GlobalTrader = GlobalTrader::new_empty(trader, free_address_deposit);
 
         require!(
             global_trader_tree.lookup_index(&global_trader) == NIL,
             ManifestError::AlreadyClaimedSeat,
             "Already claimed global trader seat",
         )?;
 
-        global_trader_tree.insert(free_address, global_trader);
+        global_trader_tree.insert(free_address_trader, global_trader);
         fixed.global_traders_root_index = global_trader_tree.get_root_index();
+        require!(
+            fixed.num_seats_claimed < MAX_GLOBAL_SEATS,
+            ManifestError::TooManyGlobalSeats,
+            "There is a strict limit on number of seats available in a global, use evict",
+        )?;
+
+        fixed.num_seats_claimed += 1;
+
+        let global_deposit: GlobalDeposit = GlobalDeposit::new_empty(trader);
+        let mut global_deposit_tree: GlobalDepositTree = GlobalDepositTree::new(
+            dynamic,
+            fixed.global_deposits_root_index,
+            fixed.global_deposits_max_index,
+        );
+        global_deposit_tree.insert(free_address_deposit, global_deposit);
+        fixed.global_deposits_root_index = global_deposit_tree.get_root_index();
+        fixed.global_deposits_max_index = global_deposit_tree.get_max_index();
+
+        Ok(())
+    }
+
+    /// Evict from the global account and steal their seat
+    pub fn evict_and_take_seat(
+        &mut self,
+        existing_trader: &Pubkey,
+        new_trader: &Pubkey,
+    ) -> ProgramResult {
+        // TODO: Make this expensive so it cannot be composed and used in attacks where bumping.
+        let DynamicAccount { fixed, dynamic } = self.borrow_mut_global();
+
+        let existing_global_trader_opt: Option<&GlobalTrader> =
+            get_global_trader(fixed, dynamic, existing_trader);
+        require!(
+            existing_global_trader_opt.is_some(),
+            ManifestError::MissingGlobal,
+            "Could not find global trader for {}",
+            existing_trader
+        )?;
+        let existing_global_trader: GlobalTrader = *existing_global_trader_opt.unwrap();
+
+        let existing_global_deposit_opt: Option<&mut GlobalDeposit> =
+            get_mut_global_deposit(fixed, dynamic, existing_trader);
+        require!(
+            existing_global_deposit_opt.is_some(),
+            ManifestError::MissingGlobal,
+            "Could not find global deposit for {}",
+            existing_trader
+        )?;
+        let existing_global_deposit: &mut GlobalDeposit = existing_global_deposit_opt.unwrap();
+
+        let existing_global_atoms_deposited: GlobalAtoms = existing_global_deposit.balance_atoms;
+        require!(
+            existing_global_atoms_deposited == GlobalAtoms::ZERO,
+            ManifestError::GlobalInsufficient,
+            "Error in emptying the existing global",
+        )?;
+
+        // Assert that the max index is the deposit index we are taking.
```

Comment:
i dont see an assert here, i also wonder why you would look up existing and then check if it is max later?